File: sdks/javascript/swr/cache.swr.ts
import { createCacheMetadata, type CacheEntry } from "../entry";
import type { CacheSingleFlight } from "../singleflight";
import type { MemoryCacheStore } from "../storage/cache.store.memory";
import type { IndexedDbCacheStore } from "../storage/cache.store.idb";
import { extractOacTtlMs } from "../ttl";
import { assertCacheEntryInvariants } from "../validation";

export interface SwrFetchResult<T> {
  data: T;
  ttlMs?: number;
  headers?: Headers | Record<string, string | undefined>;
}

export interface SwrRefreshRequest<T> {
  cacheKey: string;
  staleEntry: CacheEntry<T>;
  requestId: string;
  singleFlight: CacheSingleFlight;
  memoryStore: MemoryCacheStore<T>;
  indexedDbStore?: IndexedDbCacheStore<T>;
  fetchFromApi: (requestId: string) => Promise<SwrFetchResult<T>>;
  parity: {
    schemaVersion: string;
    specChecksum: string;
    cacheNamespace: string;
    dataVersion: string;
  };
  now?: () => number;
}

/**
 * Schedules SWR refresh as async non-blocking work with cache_key single-flight.
 */
export function scheduleSwrRefresh<T>(request: SwrRefreshRequest<T>): void {
  if (!request.requestId?.trim()) {
    throw new TypeError("SWR refresh requires the same non-empty requestId");
  }

  const now = request.now ?? (() => Date.now());

  queueMicrotask(() => {
    request.singleFlight.run(request.cacheKey, async () => {
      const api = await request.fetchFromApi(request.requestId);
      const ttlMs = api.ttlMs ?? (api.headers ? extractOacTtlMs(api.headers) : undefined);

      // OAC-authoritative missing TTL means no refresh write.
      if (ttlMs === undefined) {
        return;
      }

      const metadata = createCacheMetadata({
        source: "API",
        createdAt: now(),
        ttlMs,
        schemaVersion: request.parity.schemaVersion,
        dataVersion: request.parity.dataVersion,
        specChecksum: request.parity.specChecksum,
        cacheNamespace: request.parity.cacheNamespace,
        compressed: request.staleEntry.metadata.compressed
      });

      const refreshed: CacheEntry<T> = {
        cache_key: request.cacheKey,
        data: api.data,
        metadata
      };

      assertCacheEntryInvariants(refreshed, {
        parity: {
          schema_version: request.parity.schemaVersion,
          spec_checksum: request.parity.specChecksum,
          cache_namespace: request.parity.cacheNamespace
        },
        expectedCacheKey: request.cacheKey,
        requireNoSensitiveData: true,
        requireNoTraceInData: true
      });

      request.memoryStore.set(refreshed);
      if (request.indexedDbStore) {
        await request.indexedDbStore.set(refreshed);
      }
    }).catch(() => {
      // SWR failures are intentionally non-blocking and non-fatal.
    });
  });
}

File: sdks/flutter/lib/src/swr/cache_swr.dart
import '../entry/cache_entry.dart';
import '../singleflight/cache_single_flight.dart';
import '../storage/memory_cache_store.dart';
import '../storage/persistent_cache_store.dart';
import '../ttl/cache_ttl.dart';
import '../validation/cache_validation.dart';

class SwrFetchResult<T> {
  const SwrFetchResult({
    required this.data,
    this.ttlMs,
    this.headers,
  });

  final T data;
  final int? ttlMs;
  final Map<String, String?>? headers;
}

class SwrRefreshRequest<T> {
  const SwrRefreshRequest({
    required this.cacheKey,
    required this.staleEntry,
    required this.requestId,
    required this.singleFlight,
    required this.memoryStore,
    required this.fetchFromApi,
    required this.parity,
    this.persistentStore,
    this.now,
  });

  final String cacheKey;
  final CacheEntry<T> staleEntry;
  final String requestId;
  final CacheSingleFlight<void> singleFlight;
  final MemoryCacheStore<T> memoryStore;
  final PersistentCacheStore<T>? persistentStore;
  final Future<SwrFetchResult<T>> Function(String requestId) fetchFromApi;
  final CacheMetadataParityExpectation parity;
  final int Function()? now;
}

void scheduleSwrRefresh<T>(SwrRefreshRequest<T> request) {
  if (request.requestId.trim().isEmpty) {
    throw ArgumentError.value(request.requestId, 'requestId',
        'SWR refresh requires a non-empty requestId');
  }

  final now = request.now ?? (() => DateTime.now().millisecondsSinceEpoch);

  Future<void>.microtask(() async {
    try {
      await request.singleFlight.run(request.cacheKey, () async {
        final api = await request.fetchFromApi(request.requestId);
        final ttlMs = api.ttlMs ??
            (api.headers != null ? extractOacTtlMs(api.headers!) : null);

        if (ttlMs == null) {
          return;
        }

        final metadata = createCacheMetadata(
          source: CacheSource.api,
          createdAt: now(),
          ttlMs: ttlMs,
          schemaVersion: request.parity.schemaVersion,
          dataVersion: request.staleEntry.metadata.dataVersion,
          specChecksum: request.parity.specChecksum,
          cacheNamespace: request.parity.cacheNamespace,
          compressed: request.staleEntry.metadata.compressed,
        );

        final refreshed = CacheEntry<T>(
          cacheKey: request.cacheKey,
          data: api.data,
          metadata: metadata,
        );

        assertCacheEntryInvariants(
          refreshed,
          ValidationOptions(
            parity: request.parity,
            expectedCacheKey: request.cacheKey,
            requireNoSensitiveData: true,
            requireNoTraceInData: true,
          ),
        );

        try {
          request.memoryStore.set(refreshed);
        } catch (_) {
          // SWR memory write failures are intentionally non-blocking.
        }

        try {
          await request.persistentStore?.set(refreshed);
        } catch (_) {
          // SWR persistence failures are intentionally non-blocking.
        }
      });
    } catch (_) {
      // SWR failures are intentionally non-blocking and non-fatal.
    }
  });
}
